### PR TITLE
Avoid relying on Date.now

### PR DESCRIPTION
**Pull request by jdleesmiller**
### Background

Several issues have been opened about problems with `Date` not existing on `window`/`global`:
- #15 
- #17 
- https://github.com/testing-library/dom-testing-library/issues/194

This is in particular a problem for users of `jsdom` when using dom-testing-library in its [recommended configuration without Jest](https://testing-library.com/docs/dom-testing-library/setup#using-without-jest).

This is because, in its default configuration, `jsdom` [does not expose globals](https://github.com/jsdom/jsdom/tree/699ed6b7bea3354bf7cd7c8f174193238999cda8#executing-scripts) such as [`window.Date`](https://github.com/jsdom/jsdom/tree/699ed6b7bea3354bf7cd7c8f174193238999cda8#basic-usage)

> Important note: in the default configuration, JavaScript globals like window.Date or window.Map will not exist.

There is a workaround (https://github.com/testing-library/dom-testing-library/issues/194#issuecomment-518041038 and https://github.com/TheBrainFamily/wait-for-expect/issues/17#issuecomment-493107110), which is to change how jsdom is loaded.

There is also a PR https://github.com/TheBrainFamily/wait-for-expect/pull/20 with another attempt to solve this, but the tests are not currently passing on that branch.

See also https://github.com/testing-library/dom-testing-library/issues/194#issuecomment-467234974 from @kentcdodds .

### Proposed Solution

Rather than relying on `Date.now`, use `setTimeout` and a number of tries computed based on the timeout and interval. For example, if you want to wait 4500ms and retry every 50ms, that's (roughly) the same as saying "try 90 (=4500/50) times, once every 50ms". 

It's probably a bit less accurate, because `setTimeout` isn't exact, and the errors may accumulate, but it is probably good enough, and it avoids relying on `Date`, which might not be there in `jsdom`. And it's maybe marginally more efficient, because it doesn't have to keep checking the system clock.

### CODE REVIEW

**Review by kentcdodds — APPROVED** [review]
This looks fine to me!

**Comment by brapifra** [comment]
This still doesn't solve the problem of using fake timers as default and changing them in a particular test suite:

jest.config.json:
```object
{
  ... 
    "timers": "fake",
  ...
}
```
test.js:
```javascript
jest.useRealTimers();
describe('Test suite', () => {
   ...
});
```

setTimeout is going to be a mocked fn because of the initial configuration. **It must** be gotten every time the function is called! That was the purpose of #20

**Comment by jdleesmiller** [comment]
> setTimeout is going to be a mocked fn because of the initial configuration

Right, this PR does still require `setTimeout` to be valid when the file is loaded. This PR doesn't supersede #20, but #20 could perhaps be simplified a little bit if the code only relied on `setTimeout`, rather than both `setTimeout` and `Date`.

**Comment by brapifra** [comment]
Yeah, agreed. I'm happy to change it once you merge this!

EDIT: Nice solution btw :+1:

File: src/index.ts
Comment (brapifra) [inline src/index.ts:26]: I would probably throw an error here if `interval === 0`. Otherwise it's going to cause an infinite loop.

```suggestion
  if (interval === 0) {
    throw Error("Interval can't be 0");
  }

  const maxTries = Math.ceil(timeout / interval);
```

Comment (jdleesmiller) [reply src/index.ts:26]: Good point. The cases are
```js
Math.ceil(1 / 0) // => Infinity --- will retry forever
Math.ceil(0 / 0) // => NaN --- will also retry forever, because tries > NaN is always false
```
That is probably not what the caller intended.

An alternative to raising an error would be to clamp `interval` at 1:
```
if (interval < 1) interval = 1;
```
but that might also not be what the caller intended, because if the check is very fast, `waitForExpect` might not wait as long as it did before this change.

I am leaning toward making it an error. I've pushed up a change along the lines of what you suggest, but I figure we can also exclude negative intervals while we are at it.

This could arguably be considered a breaking change (would suggest a major version bump), because anyone who had set their interval to zero would get an error.

Comment (kentcdodds) [reply src/index.ts:26]: I say default it to 1. Nobody expects it to be so precise anyway.

**Comment by lgandecki** [comment]
Hey guys - sorry for the radio silence. I was off for a bit, getting back to the OS world now.

Thank you so much for the great work, here and in the other PR by @brapifra .
Just wanted to let you know that I'm blocking a couple hours to look at both of them closely on Friday, and I hope we can get it all merged in and released by the end of the weekend.

The adoption of this package has been really beyond the expectations - mostly caused by the success of all the *-testing-libraries by @kentcdodds and the team - so I want to be super careful and thoughtful.

File: src/index.ts
Comment (jdleesmiller) [reply src/index.ts:26]: > I say default it to 1

That's also fine by me, for the record 👍. I'm happy to make that change in this PR (or the maintainers can), but since we've now heard back from @lgandecki I'll wait for input.

Comment (lgandecki) [reply src/index.ts:26]: I agree with the default of 1 ms. 

If I understand correctly setTimeout(fn, 0) just means "next tick", while setTimeout(fn, 1) means "next tick, but not sooner than 1 ms". It's hard to imagine that "next tick" happening noticeably sooner than 1ms anyway, for these two calls to produce different results. Making software or tests rely on such unreliable assumptions (especially between dev machines and CI envs) would be very strange. 

People needing such level of precision, as Kent mentions, probably shouldn't use this tiny tool (which is built for situations that are hard to define with precision), or JS for that matter :-)

The rest looks good to me and I'm happy to merge. This is a great change and will prevent many sorts of existing and future problems.

Comment (jdleesmiller) [reply src/index.ts:26]: Great 👍 . I've pushed 329551e with that change.

**Pull request merged**
